proposals/hasCrossSiteAncestor.md

```diff
@@ -0,0 +1,122 @@
+
+Security: The hasCrossSiteAncestor boolean value does not impact security.
+
+### Exposed Sensitive Data
+No sensitive data is associated with the cross-site ancestor chain bit.
+
+### Abuse Mitigations
+A cookie may only have no cross-site ancestor when the topLevelSite in the partitionKey and the URL with which the cookie is associated with are first-party to each other. To prevent the creation of cookies that violate this, the set method will return an error if the URL and the topLevelSite are not first-party for cookies that are set with no-cross site ancestor.
+
+There are situations that can occur (such as an update to the public suffix list) which can change the if the topLevelSite and the URL are first-party to each other. To accomidate this possibility, the value for `hasCrossSiteAncestor` for the get(), getAll() and remove() methods will not be restricted. This will allow for web extensions to migrate or remove cookies that have become invalid after they have been set.
+
+### Additional Security Considerations
+None
+
+## Alternatives
+The hasCrossSiteAncestor value could be a value that is not exposed to extensions at all. Browsers that choose to include a cross-site ancestor chain bit in their partitioned cookies, could calculate the expected value of the cross-site ancestor chain bit from the URL associated with the cookie and the topLevelSite in the partitionKey. However, this could cause extensions to be unable to correctly set or get cookies (in an A1->B->A2 situation) as the browser may calculate the incorrect value for hasCrossSiteAncestor since it would not be explicitly provided by the extension.
+
+### Existing Workarounds
+To access cookies with the same `topLevelSite` but different `hasCrossSiteAncestor` values (A1->B->A2 context), developers can remove the `Partitioned` attribute from the cookie and use the Storage Access API in the context of web pages.
+
+### Open Web API
+The APIs being expanded to include the hasCrossSiteAncestor boolean are specific to extensions.
+
+## Implementation Notes
+
+### Populating hasCrossSiteAncestor when not provided
+When no value has been provided for `hasCrossSiteAncestor`, if the `domain` associated with the `cookie` is same-site to the value of the `topLevelSite`, the `hasCrossSiteAncestor` value will be set to false otherwise the value will be set to true.
+
+### APIs affected by the change and the behavior assoicated with the change:
+
+- `cookies.get()`:
+If no `hasCrossSiteAncestor` value is provided it will be populated using the algorithim described above. If a `hasCrossSiteAncestor` value is provided without a corresponding `topLevelSite` value, an error will be returned.
+
+- `cookies.getAll()`: 
+If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor.
```

Comment:
>If we want to avoid this risk, a way to do so is to filter the return values of getAll when the extension does not pass the partitionKey property, and exclude cookies whose partitionKey.hasCrossSiteAncestor do not match the default value as provided by the algorithm.

The algorithm requires both the `domain` and the `topLevelSite` to calculate the value for `hasCrossSiteAncestor`. In `cookies.getAll` if there is no `topLevelSite` (because there's no partitionKey) I'm not sure how a default value could be determined in a way that doesn't introduce a footgun. Unless this is a gap that could be covered by a new `cookies.getPartitionKey` method you [mentioned](https://github.com/w3c/webextensions/pull/581#discussion_r1703979196).

Reply:
An implementation could post-process the database results and filter those where the bit differs from the default.

The Chromium implementation already has some form of post-processing, and I can imagine such check to be added easily, at https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/extensions/api/cookies/cookies_api.cc;l=385-390;drc=3f863bca976ab92d91cc73b0e17b59b1fb4c1075.

Reply:
Given your response at https://github.com/w3c/webextensions/pull/581#discussion_r1704508645, the only valid hasCrossSiteAncestor value for unpartitioned cookies is `false`. Because of that, `getAll` can have the following behavior:

- If partitionKey is not set at all, only unpartitioned cookies are returned. This is the existing behavior and independent of hasCrossSiteAncestor.
- If partitionKey is set to an object, but hasCrossSiteAncestor is not specified, values with both true and false are returned. This is already what you are proposing.

I think that we can easily resolve this thread as follows:

```suggestion
If no value is set for hasCrossSiteAncestor cookies with both true and false values for hasCrossSiteAncestor will be returned. Otherwise, cookies will be returned that match the topLevelSite and the passed value for hasCrossSiteAncestor. When the `partitionKey` property is not specified, only unpartitioned cookies are returned. These cookies always have a `false` value for `hasCrossSiteAncestor`.
```

Reply:
SGTM, I'll commit the suggestion and resolve the conversation.